Initial commit of clang CUDA support

davidbeckingsale opened the pull request:
This commit adds support for building with Clang CUDA, but with some caveats:
1. `RAJA_ENABLE_SCAN` has been added to allow us to conditionally disable scans. We were running into issues building thrust code with clang.
2. Code compiled with `RAJA_ENABLE_NESTED` (e.g. the nested unit test) doesn't work. There is an error in the index calculations through the ForallN execution policies.

To build, just pass `-DRAJA_ENABLE_CLANG_CUDA=On -DRAJA_ENABLE_SCAN=Off` to the cmake line, specifying a new build of clang as the C and C++ compiler.

DavidPoliakoff commented:
Awesome, thanks for putting it together from those scraps, should be enough to get started with

willkill07 commented:
Awesome 👍

DavidPoliakoff commented:
👍

rhornung67 commented on cmake/SetupCompilers.cmake:
```diff
@@ -40,7 +40,9 @@
 
-set(CMAKE_CXX_STANDARD 14)
+if (NOT RAJA_ENABLE_CLANG_CUDA)
+  set(CMAKE_CXX_STANDARD 14)
```
Uhhh...how long have we been doing this? We don't want to be relying on language features that may not be available for our apps. If we allow C++14 features, then we should also have implementation alternatives that do not use anything beyond C++11 (we would probably need to add a #define to enable conditional compilation) until we verify that they work on all platforms we must support

davidbeckingsale replied:
This flag only enables C++14 if it's supported by the compiler. There is no code that relies on C++14. It silently falls back to the highest supported standard. The `RAJA` target is C++11, so enforces we use at least C++11

rhornung67 commented on cmake/modules/FindCUDA.cmake:
```diff
@@ -852,6 +852,13 @@ find_package_handle_standard_args(CUDA
+# Setup clang cuda
+if (CMAKE_CXX_COMPILER_ID MATCHES Clang 
+    AND RAJA_ENABLE_CLANG_CUDA)
+  add_definitions("-x cuda --cuda-gpu-arch=sm_35 --cuda-path=${CUDA_TOOLKIT_ROOT_DIR}")
```
I believe we have verified that everything works with sm_37 or higher.  Should we set this higher?

davidbeckingsale commented on cmake/SetupPackages.cmake:
```diff
@@ -52,6 +52,11 @@ if (RAJA_ENABLE_OPENMP)
 endif()
 
+if (RAJA_ENABLE_CLANG_CUDA)
+  set(RAJA_ENABLE_CUDA On)
```
The `RAJA_ENABLE_CLANG_CUDA` is a gross solution but it allows us to be able to build both NVCC with a clang host compiler, and CUDA with clang natively.

rhornung67 commented on include/RAJA/RAJA.hxx:
```diff
@@ -152,6 +152,10 @@
 #endif  // defined(RAJA_ENABLE_NESTED)
 
+#if defined(RAJA_ENABLE_SCAN)
+
```
Do we still have unresolved issues with Thrust?

DavidPoliakoff replied:
Yes

willkill07 replied:
The guard should only be needed for the inclusion of `exec-cuda/raja_cuda.hxx` (`#include "RAJA/exec-cuda/scan_cuda.hxx"`)

davidbeckingsale replied:
Yup, correct. If you have a chance to look at the errors thrust throws that could be helpful.

davidbeckingsale commented on include/RAJA/exec-cuda/raja_cuda.hxx:
```diff
@@ -65,6 +65,12 @@
 namespace RAJA
 {
 
+#if defined(RAJA_ENABLE_CLANG_CUDA)
```
We could actually probably remove the macro usage from the code with compiler defines, but we still need it in the CMake.

DavidPoliakoff replied:
Yeah, my hope is we can push this into a bug in the Clang bug tracker and remove it, but for now we need it. It's also broken: that's one of the things we want to discuss

davidbeckingsale commented:
I removed the RAJA_ENABLE_SCAN stuff and instead defended the scan included with `__NVCC__`, which is defined when `nvcc` is used.

davidbeckingsale commented:
I've updated this with the latest from develop, and switched over to CMake macros to centralize the switchyards that were in every CMakeLists.txt

All tests pass, apart from the GPU nested test.

rhornung67 commented on cmake/RAJAMacros.cmake:
```diff
@@ -0,0 +1,85 @@
+    if (RAJA_ENABLE_CLANG_CUDA) 
+      add_executable(${arg_NAME} ${arg_SOURCES})
+      target_compile_options(${arg_NAME} PRIVATE 
+        -x cuda --cuda-gpu-arch=sm_35 --cuda-path=${CUDA_TOOLKIT_ROOT_DIR})
```
It may be a good idea to have a CMake option for setting the compute capability (i.e., 'arch') so users can do this easily.

davidbeckingsale replied:
Yeah, I can add that.

rhornung67 commented on cmake/RAJAMacros.cmake:
```diff
@@ -0,0 +1,85 @@
+
+      add_library(${arg_NAME} ${arg_SOURCES})
+      target_compile_options(${arg_NAME} PRIVATE 
+        -x cuda --cuda-gpu-arch=sm_35 --cuda-path=${CUDA_TOOLKIT_ROOT_DIR})
```
Same comment as above...

davidbeckingsale replied:
👍

davidbeckingsale commented:
We can have a `RAJA_SM` variable that gets used. That way the user can set it to any value they want, but we will use a default value if not.

davidbeckingsale commented:
I added the `RAJA_CUDA_ARCH` variable, it's used by both nvcc and clang if set by the user when running CMake.

davidbeckingsale commented:
Note that since our 'host-config' files set the flags passed to nvcc manually, they will not be impacted by this change.
